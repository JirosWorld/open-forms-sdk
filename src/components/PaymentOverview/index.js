import PropTypes from 'prop-types';
import React from 'react';
import {FormattedMessage, defineMessage, useIntl} from 'react-intl';
import {useLocation} from 'react-router-dom';

import Body from 'components/Body';
import Card from 'components/Card';
import ErrorMessage from 'components/ErrorMessage';
import Link from 'components/Link';
import useTitle from 'hooks/useTitle';

// see openforms.payments.constants.UserAction in the backend
const USER_ACTIONS = ['accept', 'exception', 'cancel', 'unknown'];

// see openforms.payments.constants.PaymentStatus in the backend
const STATUS_MESSAGES = {
  started: defineMessage({
    description: 'payment started status',
    defaultMessage: "You've started the payment process.",
  }),
  processing: defineMessage({
    description: 'payment processing status',
    defaultMessage: 'Your payment is currently processing.',
  }),
  failed: defineMessage({
    description: 'payment failed status',
    defaultMessage:
      'The payment has failed. If you aborted the payment, please complete payment from the confirmation email.',
  }),
  completed: defineMessage({
    description: 'payment completed status',
    defaultMessage: 'Your payment has been received.',
  }),
  registered: defineMessage({
    description: 'payment registered status',
    defaultMessage: 'Your payment is received and processed.',
  }),
};

const Container = ({children}) => (
  <Card
    title={
      <FormattedMessage description="Payment overview title" defaultMessage="Payment overview" />
    }
  >
    {children}
  </Card>
);

Container.propTypes = {
  children: PropTypes.node,
};

const PaymentOverview = () => {
  const intl = useIntl();
  const {state: {status, userAction} = {}} = useLocation();

  const pageTitle = intl.formatMessage({
    description: 'Payment overview page title',
    defaultMessage: 'Payment overview',
  });
  useTitle(pageTitle);

  // display a warning if someone navigates directly to this URL.
  if (!status || !userAction) {
    return (
      <Container>
        <ErrorMessage>
          <FormattedMessage
            description="Payment overview missing state error message"
            defaultMessage="We can't display any payment information - did you maybe get here by accident?"
          />
        </ErrorMessage>
      </Container>
    );
  }

  if (!USER_ACTIONS.includes(userAction)) {
    throw new Error('Unknown payment user action');
  }

  const statusMsg = STATUS_MESSAGES[status];
  if (!statusMsg) {
    throw new Error('Unknown payment status');
  }

  let Wrapper = React.Fragment;
  if (status === 'failed') {
    Wrapper = ErrorMessage;
  }

  return (
    <Container>
      <Body component="div">
        <Wrapper>{intl.formatMessage(statusMsg)}</Wrapper>
      </Body>

      <Link to="/">
        <FormattedMessage
          description="return to form start button"
          defaultMessage="Back to form start"
        />
      </Link>
    </Container>
  );
};

PaymentOverview.propTypes = {};

export default PaymentOverview;
