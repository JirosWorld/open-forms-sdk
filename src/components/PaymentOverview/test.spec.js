import {render} from '@testing-library/react';
import messagesEN from 'i18n/compiled/en.json';
import React from 'react';
import {IntlProvider} from 'react-intl';
import {useLocation} from 'react-router-dom';
import {MemoryRouter} from 'react-router-dom';
import renderer from 'react-test-renderer';

import PaymentOverview from 'components/PaymentOverview';

/**
 * Set up mocks
 */
jest.mock('../../map/rd', () => jest.fn());

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn(),
  Link: () => <a>some link</a>,
}));

const Wrap = ({children}) => (
  <IntlProvider locale="en" messages={messagesEN}>
    <MemoryRouter>{children}</MemoryRouter>
  </IntlProvider>
);

describe('navigation without context', () => {
  it('displays a generic error message', () => {
    useLocation.mockImplementation(() => ({
      pathname: 'localhost:3000/betaaloverzicht',
    }));
    const tree = renderer
      .create(
        <Wrap>
          <PaymentOverview />
        </Wrap>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('throws on unexpected values (status)', () => {
    useLocation.mockImplementation(() => ({
      pathname: 'localhost:3000/betaaloverzicht',
      state: {
        status: 'nope',
        userAction: 'accept',
      },
    }));

    expect(() =>
      render(
        <Wrap>
          <PaymentOverview />
        </Wrap>
      )
    ).toThrow();
  });

  it('throws on unexpected values (userAction)', () => {
    useLocation.mockImplementation(() => ({
      pathname: 'localhost:3000/betaaloverzicht',
      state: {
        status: 'completed',
        userAction: 'nope',
      },
    }));

    expect(() =>
      render(
        <Wrap>
          <PaymentOverview />
        </Wrap>
      )
    ).toThrow();
  });
});

describe('on valid redirect', () => {
  it('renders the status information', () => {
    useLocation.mockImplementation(() => ({
      pathname: 'localhost:3000/betaaloverzicht',
      state: {
        status: 'completed',
        userAction: 'accept',
      },
    }));

    const tree = renderer
      .create(
        <Wrap>
          <PaymentOverview />
        </Wrap>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders an error message on failure', () => {
    useLocation.mockImplementation(() => ({
      pathname: 'localhost:3000/betaaloverzicht',
      state: {
        status: 'failed',
        userAction: 'accept',
      },
    }));

    const tree = renderer
      .create(
        <Wrap>
          <PaymentOverview />
        </Wrap>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
