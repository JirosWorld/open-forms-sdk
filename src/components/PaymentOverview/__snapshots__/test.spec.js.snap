// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation without context displays a generic error message 1`] = `
<div
  className="openforms-card"
>
  <header
    className="openforms-card__header"
  >
    <h1
      className="utrecht-heading-1 openforms-title"
    >
      Payment overview
    </h1>
  </header>
  <div
    className="openforms-card__body"
  >
     
    <div
      className="utrecht-alert utrecht-alert--error"
    >
      <div
        className="utrecht-alert__icon"
      >
        <i
          className="fa fas fa-exclamation-circle"
        />
      </div>
      <div
        className="utrecht-alert__message"
      >
        We can't display any payment information - did you maybe get here by accident?
      </div>
    </div>
     
  </div>
</div>
`;

exports[`on valid redirect renders an error message on failure 1`] = `
<div
  className="openforms-card"
>
  <header
    className="openforms-card__header"
  >
    <h1
      className="utrecht-heading-1 openforms-title"
    >
      Payment overview
    </h1>
  </header>
  <div
    className="openforms-card__body"
  >
     
    <div
      className="openforms-body"
    >
      <div
        className="utrecht-alert utrecht-alert--error"
      >
        <div
          className="utrecht-alert__icon"
        >
          <i
            className="fa fas fa-exclamation-circle"
          />
        </div>
        <div
          className="utrecht-alert__message"
        >
          The payment has failed. If you aborted the payment, please complete payment from the confirmation email.
        </div>
      </div>
    </div>
    <a
      className="utrecht-link utrecht-link--openforms"
      href="/"
      onClick={[Function]}
    >
      Back to form start
    </a>
     
  </div>
</div>
`;

exports[`on valid redirect renders the status information 1`] = `
<div
  className="openforms-card"
>
  <header
    className="openforms-card__header"
  >
    <h1
      className="utrecht-heading-1 openforms-title"
    >
      Payment overview
    </h1>
  </header>
  <div
    className="openforms-card__body"
  >
     
    <div
      className="openforms-body"
    >
      Your payment has been received.
    </div>
    <a
      className="utrecht-link utrecht-link--openforms"
      href="/"
      onClick={[Function]}
    >
      Back to form start
    </a>
     
  </div>
</div>
`;
